metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java

```diff
@@ -0,0 +1,454 @@
+                throw new RuntimeException("Unable to remove the ACL with ID " + id +
+                    " from aclsByResource");
+            }
+            if (log.isTraceEnabled()) {
+                log.trace("Removed ACL " + id + ": " + acl);
+            }
+        } catch (Throwable e) {
+            log.error("removeAcl error", e);
+            throw e;
+        }
+    }
+
+    Set<String> superUsers() {
+        return superUsers;
+    }
+
+    AuthorizationResult defaultResult() {
+        return defaultResult;
+    }
+
+    int aclCount() {
+        return aclsById.size();
+    }
+
+    static class AuthorizationResultBuilder {
+        boolean foundDeny = false;
+        boolean foundAllow = false;
+    }
+
+    /**
+     * Authorize an action based on the current set of ACLs.
+     *
+     * In order to know whether to allow or deny the action, we need to examine the ACLs
+     * that apply to it. If any DENY ACLs match, the operation is denied, no matter how
+     * many ALLOW ACLs match. If neither ALLOW nor DENY ACLs match, we return the default
+     * result. In general it makes more sense to configure the default result to be
+     * DENY, but some people (and unit tests) configure it as ALLOW.
+     */
+    AuthorizationResult authorize(AuthorizableRequestContext requestContext,
+                                  Action action) {
+        // Superusers are authorized to do anything.
+        if (superUsers.contains(requestContext.principal().toString())) {
+            if (log.isTraceEnabled()) {
+                log.trace("authorize(requestContext=" + requestContext + ", action=" + action +
+                    "): ALLOWED because " + requestContext.principal().toString() +
+                    " is a superuser");
+            }
+            return ALLOWED;
+        }
+
+        // This code relies on the ordering of StandardAcl within the NavigableMap.
+        // Entries are sorted by resource type first, then REVERSE resource name.
+        // Therefore, we can find all the applicable ACLs by starting at
+        // (resource_type, resource_name) and stepping forwards until we reach an ACL with
+        // a resource name which is not a prefix of the current one.
+        //
+        // For example, when trying to authorize a TOPIC resource named foobar, we would
+        // start at element 2, and continue on to 3 and 4 following map:
+        //
+        // 1. rs=TOPIC rn=gar pt=PREFIX
+        // 2. rs=TOPIC rn=foobar pt=PREFIX
+        // 3. rs=TOPIC rn=foob pt=LITERAL
+        // 4. rs=TOPIC rn=foo pt=PREFIX
+        // 5. rs=TOPIC rn=eeee pt=LITERAL
+        //
+        // Once we reached element 5, we would stop scanning.
+        AuthorizationResultBuilder builder = new AuthorizationResultBuilder();
+        StandardAcl exemplar = new StandardAcl(
+            action.resourcePattern().resourceType(),
+            action.resourcePattern().name(),
+            PatternType.UNKNOWN,
```

Comment:
Maybe worth mentioning somewhere in here that UNKNOWN pattern type is ordered before others since the initial search depends on that?

Reply:
Good point. I will add a comment.